Add new Analysis section to the Preferences widget

ocastejon opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://cutter.re/docs/contributing/code/getting-started.html) to this repository
- [x] I made sure to follow the project's [coding style](https://cutter.re/docs/contributing/code/development-guidelines.html)
- [ ] I've updated the [documentation](https://cutter.re/docs/user-docs.html) with the relevant information (if needed)


**Detailed description**

This PR adds a new "Analysis" section to the Preferences dialog, so that the user can modify some of the `anal.*` options and  reanalyze the file without having to do it through the console.

For the sake of completeness, I'm repeating the same considerations I put on #2228 here:

1. To keep it simple, for now I narrowed down the selected options at https://hackmd.io/@Megabeets/H1jBD3Hxv/edit to only those that have boolean values. When I have some feedback on the current code, I can add the remaining options, in this or subsequent PRs.
2. In general, I tried to keep the code as simple as possible and modify existing files as little as possible. For example, I chose to use the `Core()->getConfigb` method directly instead of `Config()->getConfig` to avoid having to modify this one.
3. Similarly, I decided to use the already existing `AnalTask` to handle the analysis. As a consequence, the title of the async task dialog is "Initial Analysis" and you can see the message "Loading the file...", which is not quite accurate:
![screenshot](https://user-images.githubusercontent.com/26287552/88467570-8d790d80-ced8-11ea-98c4-cfaff09035db.png)
However, I thought it was OK, for now at least.
4. The selected options are **not** maintained between two different sessions. However, if you think it's worth it, I can add it, as well as a "Restore defaults" button (just as in the Disassembly section).
5. I chose the icon `cog_light.svg`, since it seemed that a cog is a good representation for "Analysis" and it was already there. Let me know if that's OK, or what other icon would you think fits better. Also, I took into account that for the light theme the color of the icon changes, so added the corresponding image too.

Some other thoughts I had:

6. Now, the analyze button performs anly automated analysis (`aaa`). Maybe in the future we can add other options, or a slider just as in the initial dialog.
7. I did not add any documentation, as I didn't find anything referring to the Preferences dialog, but I can add it if needed.

**Test plan (required)**

1. Open a file without analyzing it
2. Go to Preferences -> Edit -> Analysis. The `anal.jmp.ref` and `anal.jmp.tbl` checkboxes should be the only ones checked.
3. In the console, print the value of some of the `anal.*` options and check they match the checkbox (unchecked-false/checked-true. For instance, `e anal.autoname` should return `false`.
4. Check and uncheck some boxes and verify that the returned value in the console for the corresponding configuration parameter has the correct value.
5. Close the Preferences dialog, reopen it and verify that the Analysis options are the same as before closing it.
6. Click on the analyze button. When the analysis finishes, verify that all the views have refreshed.
7. Change to the light theme and check that the icon of the Analyze section is of the same color than the rest.
![cutter-analysis-dialog](https://user-images.githubusercontent.com/26287552/88467884-4bea6180-cedc-11ea-9f35-cf10d7c999ea.gif)

**Closing issues**
Closes #2228

karliss commented:
This is only my opinion and I'd like to hear the input from others as well, don't take it as absolute truth.

I don't think most of those should be in preferences. R2 options can be roughly split into 3 categories:
* Global settings, things related user's preferred way of displaying things and some behavior
* Project settings things releated to specific session, like current architecture and other executable properties
* Configuration for specific action, options that change how action you are about to make will behave. For each run of the action you might want different values. And many of them could have been an arguments to said action but due to text based nature of r2 interface it would have been cumbersome to specify all of them each time.

In my opinion most analysis options except few like verbosity don't belong to settings. 

One of the approaches that was discussed  using approach more similar to how ghidra interface does it. Have  analysis dialog where you can configure what kind of analaysis to perform. It would be showed in two cases: during initial opening and after the executable has been opened by activating analysis action from menu. At the moment we have a set of similar settings when selecting "Advanced" analysis level in the initial options. Having the split of some randomly selected set of analysis options in initial dialog and  similar but slightly different set (without a good reason) in settings doesn't seem like good idea to me.

ocastejon commented:
Thanks for your feedback @karliss! That makes sense, if adding a new Analysis dialog was being considered. 

Of course, I can take some of these options from this section. However, I think that having the analysis options in two different places can be little bit confusing for the user. So maybe it would be worth choosing one of the two alternatives (analysis section in preferences dialog vs "standalone" analysis dialog) and putting all the options in the same place. What do you think?

I don't have any strong opinions on this at all, I'm just sharing some thoughts so we can find the best approach :smile:

ocastejon commented:
Hi @karliss, no worries, I am not discouraged at all! Actually it's quite the contrary :wink: I like to discuss the options and in the end what we all want is to find the best one, so I'm all ears!

As for the analyze button, I put it because in the list that @ITAYC0HEN shared (https://hackmd.io/@Megabeets/H1jBD3Hxv/edit) there was a line stating _Assuming there will be an “Analyze” button in this view_, so I thought that it was expected. If it's not, I'll take it out, no problem! Let me know what you think.

ITAYC0HEN commented:
I really like this PR, and especially the detailed description that shows that you paid attention to many important things. Very happy that you raised them as I didn't think of these before.

After discussion and clarifications with Karliss - we recommend the following steps:

1. Having the "Analyze program" button with a tooltip "Analyze the program using radare2's "aaa" command."
2. Add a menu item  `File -> Analyze program` to do the same. It will have the same tooltip.
3. Remove this log line (loading file) when executing the analysis from these places (keep it when running from Initial Analysis dialog)
![image](https://user-images.githubusercontent.com/20182642/89299680-af5c5800-d66f-11ea-8c41-b35e8dcbbf6f.png)

This can be a conditional print for the check that added to the right place, maybe to this `if` which is already existing. Test and let us know.
```
if (!openedFiles.size() && options.filename.length()) {
```

4. Change the title when using the dialog as "Initial Analysis" isn't right for these cases. You can either change it only for these buttons (menu item, and preferences dialog) or finding something generic for all cases such as "Analyzing program".

5. Documentation will be good. It can be very limited currently but can have a dedicated page for Preferences -> Analysis under the user docs

ocastejon commented:
Thanks for the feedback!

I have made the suggested modifications. In the following video you can see:

1. The tooltip of the "Analyze program" button
2. The new menu item `File->Analyze Program`, with the tooltip
3. The Analysis dialog without the "Loading the file" log
4. The Analysis dialog with the title "Analyzing Program" (I do not show it, but the title remains "Initial Analysis" when loading a binary and analyzing straight away)

![cutter-analysis-v2](https://user-images.githubusercontent.com/26287552/89470401-ffc0dc00-d77b-11ea-9036-688a6302f065.gif)

I moved the contents of the method `AnalOptionsWidget::analyze` to `MainWindow::on_actionAnalyze_triggered` (which was already defined, but it was empty), since it seemed that the TODO indicated it was the right place to put it. Also, then I could use the same code for the new menu item and the widget button.

I also added some documentation, let me know if that's enough.

ITAYC0HEN commented on docs/source/user-docs/preferences/analysis.rst:
```diff
@@ -0,0 +1,26 @@
+Analysis Options
+================
+
+Cutter will use the underlying radare2 `anal.*` options to analyze a binary. These options are usually 
```
I know it's a bit confusing, but unlike Markdwon, RST is using double backticks to mark inline code 

```
Like this ``inline code;`` here
```

```suggestion
Cutter will use the underlying radare2 ``anal.*`` options to analyze a binary. These options are usually 
```

Please change it in other places that single backticks where used in this PR

ITAYC0HEN commented on src/dialogs/preferences/AnalOptionsWidget.cpp:
```diff
@@ -0,0 +1,61 @@
+    Config()->setConfig(config, checkBox->isChecked());
+}
+
+/**
+ * @brief A signal to display the options in the dialog according to the current anal.* configuration
+ */
```
Same, please move it to the header file

ocastejon replied:
Done!

ocastejon commented on docs/source/user-docs/preferences/analysis.rst:
```diff
@@ -0,0 +1,26 @@
+- `anal.jmp.tbl`
+- `anal.pushret`
+- `anal.types.verbose`
+- `anal.verbose`
```
Thanks a lot for your help with the descriptions! I added them and also changed the names of the options in the UI.

ocastejon commented on src/dialogs/preferences/AnalOptionsWidget.cpp:
```diff
@@ -0,0 +1,61 @@
+ * while it's only responsible for a single independent boolean configuration eval.
+ * @param checkBox - The checkbox which is responsible for the signal
+ * @param config - the configuration string to be toggled
+ */
```
Yep, done!

ocastejon commented:
Thank you!! :smile:
